src/Razor/test/Microsoft.CodeAnalysis.Razor.Workspaces.Test/Utilities/MemoryCacheTest.cs

```diff
@@ -30,106 +32,500 @@ public async Task ConcurrentSets_DoesNotThrow()
+        var cache = new MemoryCache<string, IReadOnlyList<int>>();
+        var cacheAccessor = cache.GetTestAccessor();
+        var key = GetNewKey();
+
+        cache.Set(key, [1]);
+
+        var accessTimes = new ConcurrentBag<DateTime>();
+        var accessCount = 100;
+
+        // Concurrent access from multiple threads
+        var tasks = Enumerable.Range(0, accessCount)
+            .Select(x => Task.Run(() =>
+            {
+                _ = cache.TryGetValue(key, out _);
+
+                if (cacheAccessor.TryGetLastAccess(key, out var accessTime))
+                {
+                    accessTimes.Add(accessTime);
+                }
+            }));
+
+        await Task.WhenAll(tasks);
+
+        // Verify we got reasonable access times (no default DateTime values)
+        Assert.False(accessTimes.IsEmpty);
+        Assert.All(accessTimes, time => Assert.True(time > DateTime.MinValue));
+
+        // Verify the final access time is recent
+        Assert.True(cacheAccessor.TryGetLastAccess(key, out var finalTime));
+        Assert.True(DateTime.UtcNow - finalTime < TimeSpan.FromSeconds(1));
+    }
+
+    [Fact]
+    public async Task LRUEviction_WithConcurrentAccess_BehavesReasonably()
+    {
+        const int SizeLimit = 5;
+        var cache = new MemoryCache<string, IReadOnlyList<int>>(SizeLimit);
+        var cacheAccessor = cache.GetTestAccessor();
+
+        using var cancellationSource = new CancellationTokenSource();
+
+        // Add initial entries
+        var initialKeys = new List<string>();
+        for (var i = 0; i < SizeLimit; i++)
+        {
+            var key = $"initial-{i}";
+            initialKeys.Add(key);
+            cache.Set(key, [i]);
+        }
+
+        // Continuously access first few entries to make them "hot"
+        var hotKeys = initialKeys.Take(2).ToList();
+        var keepHotTask = Task.Run(async () =>
+        {
+            try
+            {
+                while (!cancellationSource.IsCancellationRequested)
+                {
+                    foreach (var key in hotKeys)
+                    {
+                        _ = cache.TryGetValue(key, out _);
+                    }
+
+                    await Task.Delay(1, cancellationSource.Token);
+                }
+            }
+            catch (OperationCanceledException)
+            {
+                // Expected when cancellation token is canceled
+            }
+        });
+
+        // Trigger compaction
+        await Task.Delay(10); // Let hot access pattern establish
+        cache.Set("trigger-compaction", [999]);
+
+        cancellationSource.Cancel();
+        await keepHotTask;
+
+        // Verify hot entries are more likely to survive
+        var hotSurvivalCount = hotKeys.Count(key => cache.TryGetValue(key, out _));
+        var coldSurvivalCount = initialKeys.Skip(2).Count(key => cache.TryGetValue(key, out _));
+
+        // Hot entries should have better survival rate (though this is probabilistic)
```

Comment:
>(though this is probabilistic) [](http://example.com/codeflow?start=56&length=30)

Is this probabilistic because the CTS could be cancelled before the hotKeys are requested in the cache? Can we make this not be probabilistic?